frontend/src/pages/ActivityReport/Pages/components/GoalPicker.js

```diff
@@ -13,59 +19,92 @@ const components = {
 };
 
 const GoalPicker = ({
-  control, availableGoals, selectedGoals, setValue,
+  availableGoals,
 }) => {
+  const {
+    watch, control, setValue, trigger,
+  } = useFormContext();
+  const [newGoal, updateNewGoal] = useState('');
+  const [newAvailableGoals, updateNewAvailableGoals] = useState([]);
+  const selectedGoals = watch('goals');
+
   const onRemove = (id) => {
     const newGoals = selectedGoals.filter((selectedGoal) => selectedGoal.id !== id);
+    updateNewAvailableGoals((goals) => goals.filter((goal) => goal !== id));
     setValue('goals', newGoals);
+    trigger('goals');
+  };
+
+  const onNewGoalChange = (e) => {
+    updateNewGoal(e.target.value);
   };
 
+  const addNewGoal = () => {
+    if (newGoal !== '') {
+      const goal = { id: newGoal, name: newGoal };
+      setValue('goals', [...selectedGoals, goal]);
+      trigger('goals');
+      updateNewAvailableGoals((oldGoals) => [...oldGoals, goal]);
```

Comment:
What's the advantage of this form vs. `updateNewAvailableGoals([...newAvailableGoals, goal])`?

Reply:
Since the new state depends on the old state react we need to use [functional updates](https://reactjs.org/docs/hooks-reference.html#functional-updates).